File: lesson-plans-cpls/2-IntroToGraphs/solutions/Philosophers.java
import java.util.ArrayList;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.Scanner;

public class Philosophers {

	static int solve(int N, boolean[][] E) {
	    boolean[] visited = new boolean[N];
	    int[] degrees = new int[N];
	    boolean branches = false;
	    boolean pushed = false;
	    Queue <Integer> q = new ArrayDeque<Integer>();
	    
	    for (int i = 0 ; i < N ; i++) {
	        for (int j = 0 ; j < N ; j++) {
	            if (E[j][i]) {
	                degrees[i]++;
	            }
	        }

	        if (degrees[i] == 0) {
	            if (!pushed) {
	                pushed = true;
	            }
	            else {
	                branches = true;
	            }
	            q.add(i);
	        }
	    }

	    if (!pushed) { // one big cycle
	        return 0;
	    }

	    while (!q.isEmpty()) {
	        pushed = false;
	        int a = q.poll();
	        visited[a] = true;

	        for (int i = 0 ; i < N ; i++) {
	            if (E[a][i]) {
	                if (--degrees[i] == 0) {
	                    if (!pushed) {
	                        pushed = true;
	                    }
	                    else {
	                        branches = true;
	                    }
	                    q.add(i);
	                }
	            }
	        }
	    }

	    for (int i = 0 ; i < N ; i++) {
	        if (!visited[i]) {
	            return 0; // if there sort terminated prematurely, there is a cycle
	        }
	    }

	    if (branches) {
	        return 2;
	    }

	    else {
	        return 1;
	    }
	}
	
	public static void main(String[] args) {
		Scanner s = new Scanner(System.in);
		int N = s.nextInt();
		int M = s.nextInt();



	    while ((N != 0) && (M != 0)) 
	    {
	        boolean[][] E = new boolean[N][N];

	        for (int i = 0 ; i < M ; i++) {
	            int from = s.nextInt();
	            int to = s.nextInt();
	            E[from - 1][to - 1] = true;
	        }

	        System.out.println(solve(N, E));

	        N = s.nextInt();
	        M = s.nextInt();
	    }
	  
	}


}

File: lesson-plans-cpls/2-IntroToGraphs/solutions/philosophers.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <utility>

using namespace std;

int solve(int N, bool **E) {
    bool visited[N];
    int degrees[N];
    bool branches = false;
    bool pushed = false;
    queue<int> q;
    
    for (int i = 0 ; i < N ; i++) {
        visited[i] = false;
        degrees[i] = 0;

        for (int j = 0 ; j < N ; j++) {
            if (E[j][i]) {
                degrees[i]++;
            }
        }

        if (degrees[i] == 0) {
            if (!pushed) {
                pushed = true;
            }
            else {
                branches = true;
            }
            q.push(i);
        }
    }

    if (!pushed) { // one big cycle
        return 0;
    }

    while (!q.empty()) {
        pushed = false;
        int a = q.front();
        q.pop();
        visited[a] = true;

        for (int i = 0 ; i < N ; i++) {
            if (E[a][i]) {
                if (--degrees[i] == 0) {
                    if (!pushed) {
                        pushed = true;
                    }
                    else {
                        branches = true;
                    }
                    q.push(i);
                }
            }
        }
    }

    for (int i = 0 ; i < N ; i++) {
        if (!visited[i]) {
            return 0; // if there sort terminated prematurely, there is a cycle
        }
    }

    if (branches) {
        return 2;
    }

    else {
        return 1;
    }
}

int main(int argc, char *argv[]) {
    int N, M, to, from;

    cin >> N >> M;

    while (N || M) {
        bool **E = new bool*[N];
        for (int i = 0 ; i < N ; i++) {
            E[i] = new bool[N];
            for (int j = 0 ; j < N ; j++) {
                E[i][j] = false;
            }
        }

        for (int i = 0 ; i < M ; i++) {
            cin >> from >> to;
            E[from-1][to-1] = true;
        }

        cout << solve(N, E) << '\n';

        cin >> N >> M;
    }
    return 0;
}
